### ISSUE TITLE
[FEATURE] Clarify/Improve DNS Propagation Timeouts

### ISSUE THREAD

**Issue by rbroemeling** [xenolf/lego#474]
Hi,

So I've encountered this error while testing the latest `lego` for use within our infrastructure:

```
Time limit exceeded. Last error: NS ns5.dnsmadeeasy.com. returned NXDOMAIN for <domain>
```

I understand why it occurs, and have looked far enough into it to see how I can adjust the source code to increase the propagation timeout (by modifying the provider's timeout). In my particular case, the timeout being used is for the "Dyn" provider -- and I'm hitting the timeout when the DNS propagation check happens to choose one of our secondary DNS provider's (DNS Made Easy) servers. When it happens to choose a Dyn (primary) server, the 60s allowed by the Dyn provider timeout is enough; when it chooses a DNS Made Easy (secondary) server, the 60s allowed by the Dyn provider timeout is _not_ enough.

I've also browsed through the various open issues around DNS timeouts, and see that this is a really pretty common problem.

That all said, I'm here to argue that fundamentally, timeouts can have per-provider defaults but that default should be overridable by a command-line argument. Providers having differing defaults makes sense (since generally they'll have different propagation latency characteristics), but really each individual's system is a little different (i.e., like mine; where the secondary provider servers aren't guaranteed to receive the propagation of the change within the necessary window). Allowing over-ride of this via a command-line argument in order to allow users to make adjustments for their specific situation makes sense to me.

What I would suggest is that an argument be implemented, and then the code in `dns_challenge.go` be updated to prioritize the argument over all other cases. i.e.,

```go
// Insert code here that uses the timeout from the command-line, if set.
switch provider := s.provider.(type) {
case ChallengeProviderTimeout:
	timeout, interval = provider.Timeout()
default:
	timeout, interval = 60*time.Second, 2*time.Second
}
```

I'm not completely clear on what this command-line argument should be: `--dns-timeout` already exists and is used for something different (and is symmetrical with `--http-timeout`). Perhaps `--dns-propagation-timeout` should be implemented, and then the usage/help text for `--dns-timeout` should be updated as well so that the difference between the two of them is clear?

I'm willing to take this on and provide a PR for the change, but as per the contribution guidelines I wanted to check here first to see whether such a PR would be welcomed or not.

**Comment by rbroemeling** [comment]
The more I looked at it, the more I was sure it wasn't a bunch of work; so I spent a few minutes doing an initial/WIP implementation -- see #475.

**Comment by stealthybox** [comment]
Running on kubernetes with Ingress records, I'm finding the googlecloud DNS-01 challenge to be very flaky.

I'm not sure if I'm experiencing this issue exactly, because the Google provider's timeouts seem pretty generous?
https://github.com/clio/lego/blob/dd74b99/providers/dns/googlecloud/googlecloud.go#L162-L166

Here's the relevant log snippet:
```
time="2018-04-06T21:05:49Z" level=warning msg="Endpoints not available for default/ui-ingress-ui-docs"
time="2018-04-06T21:05:57Z" level=info msg="Server configuration reloaded on :80"
time="2018-04-06T21:05:57Z" level=info msg="Server configuration reloaded on :443"
time="2018-04-06T21:05:57Z" level=info msg="Server configuration reloaded on :8080"
time="2018-04-06T21:06:59Z" level=error msg="map[ui-ingress.company.com:acme: Error 400 - urn:acme:error:connection - DNS problem: NXDOMAIN looking up TXT for _acme-challenge.ui-ingress.company.com
Error Detail:
]"
time="2018-04-06T21:06:59Z" level=error msg="Error getting ACME certificates [ui-ingress.company.com] : cannot obtain certificates map[ui-ingress.company.com:acme: Error 400 - urn:acme:error:connection - DNS problem: NXDOMAIN looking up TXT for _acme-challenge.ui-ingress.company.com
Error Detail:
]"
time="2018-04-06T21:07:59Z" level=error msg="map[ui-ingress.company.com:acme: Error 400 - urn:acme:error:connection - DNS problem: NXDOMAIN looking up TXT for _acme-challenge.ui-ingress.company.com
Error Detail:
]"
time="2018-04-06T21:07:59Z" level=error msg="Error getting ACME certificates [ui-ingress.company.com] : cannot obtain certificates map[ui-ingress.company.com:acme: Error 400 - urn:acme:error:connection - DNS problem: NXDOMAIN looking up TXT for _acme-challenge.ui-ingress.company.com
Error Detail:
]"
time="2018-04-06T21:14:40Z" level=warning msg="Endpoints not available for default/ui-ingress-ui-docs"
time="2018-04-06T21:14:40Z" level=info msg="Server configuration reloaded on :80"
time="2018-04-06T21:14:40Z" level=info msg="Server configuration reloaded on :443"
time="2018-04-06T21:14:40Z" level=info msg="Server configuration reloaded on :8080"
```

**Comment by ldez** [comment]
Closed by #634
